store/src/server/api.rs

```diff
@@ -79,14 +82,102 @@ impl api_server::Api for StoreApi {
 
         Ok(Response::new(FetchBlockHeaderByNumberResponse { block_header }))
     }
+
+    async fn sync_state(
+        &self,
+        request: tonic::Request<SyncStateRequest>,
+    ) -> Result<Response<SyncStateResponse>, Status> {
+        let request = request.into_inner();
+
+        let notes = self
+            .db
+            .notes_since_block_by_tag(&request.note_tags, request.block_num)
+            .await
+            .map_err(|err| Status::internal(format!("{:?}", err)))?;
+
+        let (block_header, chain_tip) = if !notes.is_empty() {
+            let block_header = self
+                .db
+                .get_block_header(Some(notes[0].block_num))
+                .await
+                .map_err(|err| Status::internal(format!("{:?}", err)))?
+                .ok_or(Status::internal("Block db is empty"))?;
+            let tip = self
+                .db
+                .get_block_header(None)
+                .await
+                .map_err(|err| Status::internal(format!("{:?}", err)))?
+                .ok_or(Status::internal("Block db is empty"))?;
+
+            (block_header, tip.block_num)
+        } else {
+            let block_header = self
+                .db
+                .get_block_header(None)
+                .await
+                .map_err(|err| Status::internal(format!("{:?}", err)))?
+                .ok_or(Status::internal("Block db is empty"))?;
+
+            let block_num = block_header.block_num;
+            (block_header, block_num)
+        };
+
+        let account_ids: Vec<u64> = request.account_ids.iter().map(|e| e.id).collect();
+        let accounts = self
+            .db
+            .get_account_hash_by_block_range(
+                request.block_num,
+                block_header.block_num,
+                &account_ids,
+            )
+            .await
+            .map_err(|err| Status::internal(format!("{:?}", err)))?;
+
+        let accounts: Vec<_> = accounts
+            .into_iter()
+            .map(|(id, account_hash, block_num)| AccountHashUpdate {
+                account_id: Some(AccountId { id }),
+                account_hash: Some(account_hash),
+                block_num,
+            })
+            .collect();
+
+        let nullifiers = self
+            .db
+            .get_nullifiers_by_block_range(
+                request.block_num,
+                block_header.block_num,
+                &request.nullifiers,
+            )
+            .await
+            .map_err(|err| Status::internal(format!("{:?}", err)))?;
+
+        let nullifiers: Vec<_> = nullifiers
+            .into_iter()
+            .map(|(nullifier, block_num)| NullifierUpdate {
+                nullifier: Some(nullifier.into()),
+                block_num,
+            })
+            .collect();
+
+        Ok(Response::new(SyncStateResponse {
+            chain_tip,
+            block_header: Some(block_header),
+            mmr_delta: todo!(),
+            block_path: todo!(),
```

Comment:
The `mmr` only supports constructing a delta against its latest version. So this can't yet be done, I'm looking into a fix for this to add in the `crypto` repo.

Reply:
opened: https://github.com/0xPolygonMiden/crypto/pull/205 https://github.com/0xPolygonMiden/crypto/pull/206 https://github.com/0xPolygonMiden/crypto/pull/207